### DISCUSSION TITLE
Are B-spline kernels better than Peskin kernels?

### DISCUSSION THREAD

**Discussion by shaoyaoqian** [IBAMR/IBAMR#1758] [General]
I have read some recent literature and found that the regularized delta function can be chosen as a B-spline function. The B-spline function does not satisfy the even-odd condition and does not fully meet translational invariance, but it seems to perform better?

**Comment by boyceg** [comment]
We have two papers that look at this:

* https://doi.org/10.1016/j.jcp.2022.111042 (also available via https://arxiv.org/abs/2105.14536)
* https://arxiv.org/abs/2408.08280

This second paper includes some numerical and analytical evidence that using different kernels in different coordinate directions (I've been calling these "composite" kernels) provides an IB method with some nice properties. So far, the only composite kernels that we have worked with are B-spline kernels.

At least at this point, I don't think that we have any results that suggest that one kernel construction is always better than another. However, a couple of points:
1. No compact kernel gives perfect translation invariance.
2. The even-odd condition is mostly motivated by pressure instabilities in collocated discretizations of the incompressible Navier-Stokes equations that are not present in other discretizations --- we like to use staggered-grid (MAC) schemes that do not suffer from these instabilities.
3. Yoichiro Mori has work that shows that the even-odd condition is an example of a smoothing-order condition --- see https://doi.org/10.1137/110836699. It could be interesting to see how those compare.
4. However, at least in our empirical tests, kernels without the even-odd condition are usually more accurate than the corresponding function without that condition.

My guess is that the "best" kernel function will be context dependent. However, one thing that is clear is that different kernels do give different accuracies --- it is probably worthwhile trying a few out for a particular application or class of applications!

**Reply by shaoyaoqian** [reply]
Thank you! 

You’ve provided exactly the information I was looking for. I now have a clearer understanding of the concept.

I have read the two papers you provide. Besides, I have read the code from https://github.com/stochasticHydroTools/IBMethod based on the paper http://arxiv.org/abs/1505.07529 and found the comments regarding the invariance of B-splines. 

 Thanks again!

**Comment by shaoyaoqian** [comment]
Hi, Boyce. 

I am very interested in the IB method and I am writing simple code to verify my ideas about IB kernels. I found that high order B-spline kernel won’t produce interpolating operator with accuracy higher than second order. 

For example, there is a 1D funciton $f(x)$, the interpolating operator can be expressed as

$$
F(X_k)=\sum_{i} f(x_i)\delta_h(x_i-\chi(X_k))h,
$$

where $\delta_h$ is a B-spline kernel. I noticed that 

$$
x^*=\chi(X_k),
$$

$$
|F(X_k)- f(x^*)|=O(h^2).
$$

 Increase the order of B-spline, $e=|F(X_k)- f(x^*)|$ won't decrease.



I am confused, is it the problem of my code? 



I have an unverified idea. Let $\phi_i(x)=\delta_h(x_i-x)$, project $f(x)$ into B-spline function space and get the coefficient $f_i^*$.

 $f(x)$ can be approximated with $\sum f\_i^*\phi\_i(x)$ but

$$
f(x\_i)\neq f\_i^*.
$$

Finally,

$$
F(X_k)=\sum_{i} f_i^*\delta(x_i-\chi(X_k))h.
$$

I think $F(X_k)$ might be more accurate and I am trying this.

**Reply by boyceg** [reply]
A couple quick comments.

I don't think that the cardinal B-spline functions satisfy high-order discrete moment conditions (which are equivalent to higher-order accuracy conditions when interpolating smooth fields); see Table 1 in this paper: https://doi.org/10.1016/j.jcp.2022.111042. However, I don't think that is really a problem in practice. For most practical IB computations, the velocity is $$C^0$$, and pressure field is discontinuous. Consequently, imposing conditions on the construction of $$\delta_h$$ that leads to higher-order accuracy for ***smooth*** fields does not necessarily improve the accuracy of the method in practice --- without some other corrections, the reconstructed velocity will only converge at $$O(h)$$.

If you want to effectively use regularized delta functions that satisfy higher-order interpolation conditions, one possible route is to try to "remove" the rough parts of the field that is being computed. In general, if you can decompose the field as $$u = u^\text{smooth} + u^\text{rough}$$, and if it is easy to evaluate $$u^\text{rough}$$ analytically, then you could try to use higher-order interpolation only on $$u^\text{smooth}$$. Here is a paper that does something like this for the pressure but not the velocity: https://doi.org/10.1016/j.jcp.2019.109217.

It sounds like you are also thinking about using the regularized delta functions as basis functions for the force and velocity. I think that this is a good idea.

**Reply by shaoyaoqian** [reply]
Thank you, Boyce !

**Comment by shaoyaoqian** [comment]
Hi Boyce!
I have a finding. I am considering whether the local divergence free kernels can be used in the IFED method, which is introduced in the paper.

>  **B.E. Griffith, X.Y. Luo, Hybrid finite difference/finite element immersed boundary method, Int. J. Numer. Methods Biomed. Eng. 33 (11) (2017) e2888.**

To ensure the adjoint of interpolating operator and spreading operator, there need a projecting step after the velocity is interpolated into the Gauss quadrature points of the solid. In my practice, this will break the divergence free property of the velocity.

**Reply by boyceg** [reply]
We are actively looking into this and should have a preprint ready in a month or two.

**Reply by shaoyaoqian** [reply]
That's awesome, looking forward to reading it!

**Reply by shaoyaoqian** [reply]
Hi Boyce!

Sorry for bothering you again. Has the preprint been uploaded?

**Reply by boyceg** [reply]
The study is still in progress.

**Reply by shaoyaoqian** [reply]
Hi Boyce!

After a long wait, I've finally had the chance to read your paper!

It's highly relevant to the ones you've shared with me previously and I studied the method detailed in the paper quite well, but I do have a question: Is it necessary to calculate $w_m$ explicitly? It seems unnecessary to calculate the nodal quadrature weight in the fully nodal quadrature method.

<img width="1043" alt="image" src="https://github.com/user-attachments/assets/34f7cb02-7ae6-4f18-ab60-aa0af6b90518" />

**Reply by boyceg** [reply]
You are correct that the numerical values of these weights ultimately drop out, and their specific values do not impact the computation. We include this in the paper because this is how we actually do things in practice — we try to use sensible quadrature weights.

**Reply by boyceg** [reply]
Now, it is true that we could re-write the discretized formulation to avoid using these nodal quadrature weights, but then I think that the linkage between the continuum and discrete equations becomes less clear. (Maybe I'm wrong.)

One concrete advantage of using the approach described in the paper in an implementation is that it makes it easy to switch between "consistent" and "lumped" mass matrices.

**Reply by shaoyaoqian** [reply]
Thank you !

Read the paper again, I just noticed the explanation after equation (19).

**Comment by shaoyaoqian** [comment]
Hello, Boyce!

I am back again. I am investigating the adjoint property of the interpolating operator and spreading operator. I find that the projection step only suitable for first order finite element. But,  I might be wrong.

> **B.E. Griffith, X.Y. Luo, Hybrid finite difference/finite element immersed boundary method, Int. J. Numer. Methods Biomed. Eng. 33 (11) (2017) e2888.**

I did not use Gauss quadrature points, I just use the node of the mesh and the rectangular numerical integration formula.

Let $V_h = \text{span}\\{\phi_i, \; 0 \le i \le N\\}$ represent the finite element space, with $\phi_i$ being the Lagrange interpolation basis functions, and $N$ being the number of interpolation nodes, which corresponds to the dimension of the finite element space. Let $f$ be an arbitrary function, and define $[f]$ as the vector composed of the values at the mesh nodes, where $[f]_i$ denotes the $i$-th component.

For the adjoint of the operators, I do the following steps.
Given a function $b$, for any function $a$, there exists a function $b^*$  such that the following equation (1) holds:

$$
\sum_{i} \sum_{j} [b^*]_i [a]_j (\phi_i, \phi_j) = \sum_j a_j b_j w_j \tag{1}
$$

where $w_j$ is the area accounting for the $i$-th node. Thus, we have:

$$
\sum_j [a]_j \sum_i [b^*]_i (\phi_i, \phi_j) = \sum_j a_j b_j w_j \tag{2}
$$

Due to the arbitrariness of $a$, it follows that:

$$
\sum_i [b^*]_i (\phi_i, \phi_j) = b_j w_j, \quad \forall 0 \le j \le N \tag{3}
$$

Finally, we calculate $b^*$.

I have written some code the verify the idea of projection. Here is my conclusion, when first order element is used, $b^*$ is similar to $b$. But when finite elements larger than 2nd order is used, $b^\*$ is much larger than $b$.

**Reply by boyceg** [reply]
This paper discusses how to do force spreading and velocity interpolation just from the nodes: https://doi.org/10.1016/j.jcp.2022.111890

**Reply by boyceg** [reply]
The basic idea is that you need to pick some inner product to define the notion of adjoint.

If you treat the integrals "exactly" or use high-order quadrature, then you wind up needing to perform a standard L2 projection in the velocity interpolation step.

If you use a nodal quadrature rule to define the inner product, then you get a scheme that only involves spreading forces from the nodes and interpolating velocities to the nodes. What we discuss in the paper that I linked to above is that, when you use a nodal quadrature rule in "just the right way", the choice of quadrature weights drops out.
